pwqcheck.1
.\" Copyright (c) 2009 Dmitry V. Levin
.\" All rights reserved.
.\" Copyright (c) 2000-2003,2005,2008,2010,2019,2020,2025 Solar Designer
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd November 23, 2025
.Dt PWQCHECK 1
.Os "Openwall Project"
.Sh NAME
.Nm pwqcheck
.Nd Check passphrase quality
.Sh SYNOPSIS
.Nm Op Ar options
.Sh DESCRIPTION
The
.Nm
program checks passphrase quality using the libpasswdqc library.
By default, it expects to read 3 lines from standard input:
.Pp
.Bl -item -compact -offset indent
.It
first line is a new password,
.It
second line is an old password, and
.It
third line is either an existing account name or a
.Xr passwd 5
entry.
.El
.Pp
There are a number of supported options, which can be used to control the
.Nm
behavior.
.Pp
.Nm
prints
.Ar OK
on success.  Scripts invoking
.Nm
are suggested to check for both a zero exit status and the
.Ar OK
line.
.Sh OPTIONS
.Bl -tag -width Ds
.Sm off
.It Xo
.Cm min No =
.Ar N0 , N1 , N2 , N3 , N4
.Xc
.Sm on
.Pq default: min=disabled,24,11,8,8
The minimum allowed password lengths for different kinds of
passwords/passphrases.
The keyword
.Cm disabled
can be used to
disallow passwords of a given kind regardless of their length.
Each subsequent number is required to be no larger than the preceding
one.
.Pp
.Ar N0
is used for passwords consisting of characters from one character
class only.
The character classes are: digits, lower-case letters, upper-case
letters, and other characters.
There is also a special class for
.No non- Ns Tn ASCII
characters, which could not be classified, but are assumed to be non-digits.
.Pp
.Ar N1
is used for passwords consisting of characters from two character
classes that do not meet the requirements for a passphrase.
.Pp
.Ar N2
is used for passphrases.
Note that besides meeting this length requirement,
a passphrase must also consist of a sufficient number of words (see the
.Cm passphrase
option below).
.Pp
.Ar N3
and
.Ar N4
are used for passwords consisting of characters from three
and four character classes, respectively.
.Pp
When calculating the number of character classes, upper-case letters
used as the first character and digits used as the last character of a
password are not counted.
.Pp
In addition to being sufficiently long, passwords are required to
contain enough different characters for the character classes and
the minimum length they have been checked against.
.Pp
.It Cm max Ns = Ns Ar N
.Pq default: Cm max Ns = Ns 72
The maximum allowed password length.
This can be used to prevent users from setting passwords that may be
too long for some system services.
The value 8 is treated specially: if
.Cm max
is set to 8, passwords longer than 8 characters will not be rejected,
but will be truncated to 8 characters for the strength checks and the
user will be warned.
This is to be used with the traditional DES-based password hashes,
which truncate the password at 8 characters.
.Pp
It is important that you do set
.Cm max Ns = Ns 8
if you are using the traditional
hashes, or some weak passwords will pass the checks.
.It Cm passphrase Ns = Ns Ar N
.Pq default: Cm passphrase Ns = Ns 3
The number of words required for a passphrase.
.It Cm match Ns = Ns Ar N
.Pq default: Cm match Ns = Ns 4
The length of common substring required to conclude that a password is
at least partially based on information found in a character string,
or 0 to disable the substring search.
Note that the password will not be rejected once a weak substring is
found; it will instead be subjected to the usual strength requirements
with the weak substring partially discounted.
.Pp
The substring search is case-insensitive and is able to detect and
remove a common substring spelled backwards.
.It Xo
.Sm off
.Cm similar No = Cm permit | deny
.Sm on
.Xc
.Pq default: Cm similar Ns = Ns Cm deny
Whether a new password is allowed to be similar to the old one.
The passwords are considered to be similar when there is a sufficiently
long common substring and the new password with the substring partially
discounted would be weak.
.It Cm wordlist Ns = Ns Ar FILE
Deny passwords that are based on lines of the tiny external text
.Ar FILE ,
which can reasonably include common passwords, dictionary words in a
local language other than English, and longer English words.
.Pp
passwdqc includes built-in lists of a few thousand common English words
mostly of lengths from 3 to 6, a handful of common character sequences
such as keyboard walks, and many common passwords of length 7 and above.
Any word list possibly specified with this option is used in addition to
the lists of English words and character sequences, but replaces the
list of common passwords.
.Pp
Substring matching and discounting will be used if the
.Cm match
setting
above is non-zero.  Please note that this isn't to be used with large
wordlists for performance reasons.
.It Cm denylist Ns = Ns Ar FILE
Deny passwords or passphrases directly appearing in the tiny external text
.Ar FILE .
That file can reasonably be e.g. a list of common passwords if
only a relaxed policy is desired and stricter checks are thus disabled
(using their separate options).  Such policy would only be somewhat
effective against online/remote attacks, but not against offline attacks
on hashed passwords.
.It Cm filter Ns = Ns Ar FILE
Deny passwords or passphrases directly appearing in a maybe huge binary
filter
.Ar FILE
created with pwqfilter.  This is very efficient, needing at
most two random disk reads per query.  A filter created from millions of
leaked passwords can reasonably be used on top of passwdqc's other
checks to further reduce the number of passing yet weak passwords
without causing unreasonable inconvenience (as e.g. higher minimum
lengths and character set requirements could).
.It Cm config Ns = Ns Ar FILE
Load config
.Ar FILE
in the
.Cm passwdqc.conf
format.  This file may define any options described in
.Xr passwdqc.conf 5 , but only the
.Cm min ,
.Cm max ,
.Cm passphrase ,
.Cm match Ns ,
and
.Cm config
options are honored by
.Nm .
.It Cm -1
Read just 1 line (new passphrase).
This is needed to use
.Nm
as the passwordcheck program on OpenBSD - e.g., with
":passwordcheck=/usr/bin/pwqcheck \-1:\\"
(without the quotes, but with the trailing backslash)
in the "default" section in
.Cm /etc/login.conf .
.It Cm -2
Read just 2 lines (new and old passphrases).
.It Cm --multi
Check multiple passphrases (until EOF).
This option may be used on its own or along with the
.Cm -1
or
.Cm -2
options.
.Nm
will read 1, 2, or 3 lines and will output one line per passphrase to check.
The lines will start with either
.Ar OK
or a message explaining why the passphrase did not pass the checks,
followed by a colon and a space, and finally followed by the passphrase.
The explanatory message is guaranteed to not include a colon.
With this option, the exit status of
.Nm
depends solely on whether there were any errors preventing the strength of
passphrases from being fully checked or not.
A primary use for this option is to test different policies and/or different
versions of passwdqc on large passphrase lists.
.It Cm --version
Output
.Nm
program version and exit.
.It Cm -h , --help
Output
.Nm
help text and exit.
.El
.Sh EXIT STATUS
.Nm
exits with non-zero status when it encounters invalid config file,
invalid option, invalid parameter value, invalid data in standard input,
and in any case when it fails to check passphrase strength.
Without the
.Cm --multi
option,
.Nm
also exits with non-zero status when it detects a weak passphrase.
.Sh FILES
.Pa /etc/passwdqc.conf
(not read unless this suggested file location is specified with the
.Cm config=/etc/passwdqc.conf
option).
.Sh SEE ALSO
.Xr pwqgen 1 ,
.Xr libpasswdqc 3 ,
.Xr passwd 5 ,
.Xr passwdqc.conf 5 ,
.Xr pam_passwdqc 8 .
.Pp
https://www.openwall.com/passwdqc/
.Sh AUTHORS
The pam_passwdqc module was written for Openwall GNU/*/Linux by Solar Designer.
The
.Nm
program was originally written for ALT GNU/*/Linux by Dmitry V. Levin,
indirectly reusing code from pam_passwdqc (via libpasswdqc).
This manual page (derived from the pam_passwdqc documentation)
was written for Openwall GNU/*/Linux by Dmitry V. Levin.


CHANGES
	Significant changes between 2.0.2 and 2.0.3.

Added Cygwin support.

Added pkg-config file.

Changed enforce=users to support "chpasswd" PAM service in addition to
traditionally supported "passwd".


	Significant changes between 2.0.1 and 2.0.2.

Improved pam_passwdqc's auto-generated policy descriptions further, so
that lines are wrapped at a more consistent length.

Added the libpasswdqc(3) manual page and links to it for all functions
documented in there.

Added scripts to support Continuous Integration on GitHub (included in
the git repository, but excluded from release tarballs).


	Significant changes between 2.0.0 and 2.0.1.

Improved pam_passwdqc's auto-generated policy descriptions, which were
slightly misformatted since the introduction of i18n support in 1.4.0.
Now they not only look prettier, but also make it clearer that the
mentioned lengths are merely the minimums and not the recommended ones.

Updated Russian translation for consistency with the above and to cover
messages added in 1.9.0+.

Increased maximum size of randomly-generated passphrases to 136 bits.
This was already the limit in the underlying API, but the tools' limit
was set to 85.  This increase is to allow for a wider variety of use
cases for the tools.

In the Makefile, use CPPFLAGS and LDFLAGS consistently to be friendlier
to packaging by distros.

Added this file CHANGES based on two latest release announcements, and
started to maintain it.


	Significant changes between 1.9.0 and 2.0.0.

Introduce and use passwdqc_params_free().  This is a minor addition to
the libpasswdqc API related to the addition of external files support.


	Significant changes between 1.5.0 and 1.9.0.

Added support for external wordlist, denylist, and binary filter.  With
these, passwdqc can be configured to deny passwords and passphrases that
are based on lines of a tiny external text file (the "wordlist" option),
directly appear in a tiny external text file (the "denylist" option),
or/and directly appear in a maybe huge binary filter file (the "filter"
option).  While usage of larger external text files is inefficient, the
binary filters are very efficient.

The binary filters can be created and otherwise managed with the newly
added pwqfilter(1) program.  It can create a binary filter from a list
of plaintexts or from MD4 or NTLM hashes.  The latter are supported in a
way that enables importing of HIBP (Pwned Passwords) database revisions
into passwdqc binary filters.  pwqfilter works on arbitrary plain text
strings or hex-encoded hashes, and it can also be reused in lieu of
grep(1) for many purposes, even unrelated to passphrases and security.

Merged changes needed for building with Visual Studio on Windows.  This
includes a refactoring of the random passphrase generator code to make
it shared between platforms.


	Significant changes between 1.4.1 and 1.5.0.

Updated the included wordlist to avoid some inappropriate words in
randomly generated passphrases while not removing any words from the
"word-based" check, and also to have plenty of extra words for
subsequent removal of more words that might be considered inappropriate
from the initial 4096 that are used for randomly generated passphrases.
Most of the added words came from EFF Diceware, BIP-0039, and our own
processing of Project Gutenberg Australia books.


	Significant changes between 1.4.0 and 1.4.1.

Set default for "max" to 72 (was 40).  The previous setting was based on
a reading of RFC 1939, which in practice did not matter.  The new one is
based on bcrypt's truncation at 72, which actually still matters.

Documented "similar" in pwqcheck(1) help message and manual page.  This
is a setting that was supported before and documented for other passwdqc
components before, but was apparently erroneously omitted from here.


	Significant changes between 1.3.2 and 1.4.0.

Implemented i18n support in pam_passwdqc, contributed by Oleg Solovyov,
Andrey Cherepanov, and Dmitry V. Levin.  The i18n support is off by
default, it can be enabled if Linux-PAM is built using --enable-nls
configure option.

Implemented audit support in pam_passwdqc, contributed by Oleg Solovyov
and Dmitry V. Levin.  The audit support is off by default, it can be
enabled if Linux-PAM is built using --enable-audit configure option.

Both of these optional new features had been introduced and are enabled
in ALT Linux distributions, so this version is effectively upstreaming
the changes from there.


	Significant changes between 1.3.1 and 1.3.2.

Compatibility for building with newer versions of glibc, where we now
have to define _DEFAULT_SOURCE for our use of crypt(3).  The problem was
identified and this change tested by Dmitry V. Levin.

Clarified in the man pages that /etc/passwdqc.conf is not read unless
this suggested file location is specified with the config= option.

Clarified the OpenBSD configuration example.

Escape the minus sign in the OpenBSD configuration example to make the
man page linter happy, patch by Jackson Doak via Unit 193.


pwqcheck.c
/*
 * Copyright (c) 2008,2009 by Dmitry V. Levin
 * Copyright (c) 2010,2016,2021 by Solar Designer
 * See LICENSE
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "passwdqc.h"

static void clean(char *dst, size_t size)
{
	if (!dst)
		return;
	_passwdqc_memzero(dst, size);
	free(dst);
}

static char *read_line(size_t size, int eof_ok)
{
	char *p, *buf = malloc(size + 1);

	if (!buf) {
		fprintf(stderr, "pwqcheck: Memory allocation failed.\n");
		return NULL;
	}

	if (!fgets(buf, size + 1, stdin)) {
		clean(buf, size + 1);
		if (!eof_ok || !feof(stdin) || ferror(stdin))
			fprintf(stderr,
			    "pwqcheck: Error reading standard input.\n");
		return NULL;
	}

	if (strlen(buf) >= size) {
		clean(buf, size + 1);
		fprintf(stderr, "pwqcheck: Line too long.\n");
		return NULL;
	}

	if ((p = strpbrk(buf, "\r\n")))
		*p = '\0';

	return buf;
}

static char *extract_string(char **stringp)
{
	char *token = *stringp, *colon;

	if (!token)
		return "";

	colon = strchr(token, ':');
	if (colon) {
		*colon = '\0';
		*stringp = colon + 1;
	} else
		*stringp = NULL;

	return token;
}

static struct passwd *parse_pwline(char *line, struct passwd *pw)
{
	if (!strchr(line, ':')) {
#ifdef _MSC_VER
		memset(pw, 0, sizeof(*pw));
		pw->pw_name = line;
#else
		struct passwd *p = getpwnam(line);
		endpwent();
		if (!p) {
			fprintf(stderr, "pwqcheck: User not found.\n");
			return NULL;
		}
		if (p->pw_passwd)
			_passwdqc_memzero(p->pw_passwd, strlen(p->pw_passwd));
		memcpy(pw, p, sizeof(*pw));
#endif
	} else {
		memset(pw, 0, sizeof(*pw));
		pw->pw_name = extract_string(&line);
		pw->pw_passwd = extract_string(&line);
		extract_string(&line); /* uid */
		extract_string(&line); /* gid */
		pw->pw_gecos = extract_string(&line);
		pw->pw_dir = extract_string(&line);
		pw->pw_shell = line ? line : "";
		if (!*pw->pw_name || !*pw->pw_dir) {
			fprintf(stderr, "pwqcheck: Invalid passwd entry.\n");
			return NULL;
		}
	}
	return pw;
}

static void
print_help(void)
{
	puts("Check passphrase quality.\n"
	    "\nFor each passphrase to check, pwqcheck reads up to 3 lines from standard input:\n"
	    "  first line is a new passphrase,\n"
	    "  second line is an old passphrase, and\n"
	    "  third line is either an existing account name or a passwd entry.\n"
	    "\nUsage: pwqcheck [options]\n"
	    "\nValid options are:\n"
	    "  min=N0,N1,N2,N3,N4\n"
	    "       set minimum allowed lengths for different kinds of passphrases;\n"
	    "  max=N\n"
	    "       set maximum allowed passphrase length;\n"
	    "  passphrase=N\n"
	    "       set number of words required for a passphrase;\n"
	    "  match=N\n"
	    "       set length of common substring in substring check;\n"
	    "  similar=permit|deny\n"
	    "       whether a new passphrase is allowed to be similar to the old one;\n"
	    "  wordlist=FILE\n"
	    "       deny passwords that are based on lines of a tiny external text file;\n"
	    "  denylist=FILE\n"
	    "       deny passphrases directly appearing in a tiny external text file;\n"
	    "  filter=FILE\n"
	    "       deny passphrases directly appearing in a maybe huge binary filter file;\n"
	    "  config=FILE\n"
	    "       load config FILE in passwdqc.conf format;\n"
	    "  -1\n"
	    "       read just 1 line (new passphrase);\n"
	    "  -2\n"
	    "       read just 2 lines (new and old passphrases);\n"
	    "  --multi\n"
	    "       check multiple passphrases (until EOF);\n"
	    "  --version\n"
	    "       print program version and exit;\n"
	    "  -h or --help\n"
	    "       print this help text and exit.");
}

int main(int argc, const char **argv)
{
	passwdqc_params_t params;
	const char *check_reason;
	char *parse_reason, *newpass, *oldpass, *pwline;
	struct passwd pwbuf, *pw;
	int lines_to_read = 3, multi = 0;
	size_t size = 8192;
	int rc = 1;

	while (argc > 1 && argv[1][0] == '-') {
		const char *arg = argv[1];

		if (!strcmp("-h", arg) || !strcmp("--help", arg)) {
			print_help();
			return 0;
		}

		if (!strcmp("--version", arg)) {
			printf("pwqcheck version %s\n", PASSWDQC_VERSION);
			return 0;
		}

		if ((arg[1] == '1' || arg[1] == '2') && !arg[2]) {
			lines_to_read = arg[1] - '0';
			goto next_arg;
		}

		if (!strcmp("--multi", arg)) {
			multi = 1;
			goto next_arg;
		}

		break;

next_arg:
		argc--; argv++;
	}

	passwdqc_params_reset(&params);
	if (argc > 1 &&
	    passwdqc_params_parse(&params, &parse_reason, argc - 1,
		argv + 1)) {
		fprintf(stderr, "pwqcheck: %s\n",
		    (parse_reason ? parse_reason : "Out of memory"));
		free(parse_reason);
		return rc;
	}

	if ((size_t)params.qc.max + 1 > size)
		size = (size_t)params.qc.max + 1;

next_pass:
	oldpass = pwline = NULL; pw = NULL;
	if (!(newpass = read_line(size, multi))) {
		if (multi && feof(stdin) && !ferror(stdin) &&
		    fflush(stdout) >= 0)
			rc = 0;
		goto done;
	}
	if (lines_to_read >= 2 && !(oldpass = read_line(size, 0)))
		goto done;
	if (lines_to_read >= 3 && (!(pwline = read_line(size, 0)) ||
	    !parse_pwline(pwline, pw = &pwbuf)))
		goto done;

	check_reason = passwdqc_check(&params.qc, newpass, oldpass, pw);
	if (!check_reason) {
		if (multi)
			printf("OK: %s\n", newpass);
		else if (puts("OK") >= 0 && fflush(stdout) >= 0)
			rc = 0;
		goto cleanup;
	}
	if (multi)
		printf("Bad passphrase (%s): %s\n", check_reason, newpass);
	else
		printf("Bad passphrase (%s)\n", check_reason);

cleanup:
	_passwdqc_memzero(&pwbuf, sizeof(pwbuf));
	clean(pwline, size);
	clean(oldpass, size);
	clean(newpass, size);

	if (multi)
		goto next_pass;

	passwdqc_params_free(&params);

	return rc;

done:
	multi = 0;
	goto cleanup;
}
